pages/glossary/mechanism.md

```diff
@@ -0,0 +1,12 @@
+---
+title: Mechanism
+key: mechanism
+unambiguous: true
+objective: false
+---
+
+A mechanism is a set of [HTML elements][] and [scripting][] in an [HTML document][] that allow a user to achieve an objective. 
```

Comment:
I like that. I think this needs an assumption (to be copied in rules) in the line of "mechanism is implemented by content" (the rule currently already have that assumption).

---
I am not so sure about the "allow a user to achieve an objective".

A mechanism could actually be used to **prevent** a user from achieving an objective (e.g. block unauthorised access).

---
It is also not completely clear what "objective" is (high or low level?) and how much of grouping that implies in the mechanism. My objective (as a user) is not really to expand a menu, sort a table, … It's more like accessing information. Of course, sorting a table is an important step to access the information (in a meaningful way). But then, would the mechanism be the table and all its functionalities (sorting, pagination, maybe filtering, …) or even the full web page? Or would it be just the sorting button (+script) of a single column (and sorting another column is a different mechanism, …)

It's of course hard to speak about the intention ("objective") of the end user… But I feel this is usually larger high level while we want the mechanism (the set of stuff) be as small as possible. Maybe having some sort of minimality condition on the set (e.g. we want "button+script" together, but not "two buttons+script" even if they call the same function in their `onclick`). But that will also be a bit tricky to define precisely.

Maybe I'm just reading too much into this, and this is just subjectivity of the definition…

Reply:
1. I've added an assumption in line with your suggestion.1. I've added an assumption in line with your suggestion.
2. "Blocking unauthorised access" is achieving an objective, not preventing it ;-) I don't think we need to make that distinction because, probably, we could write every objective in a way that we are achieving it and in a way that we are preventing it. It would be interesting to hear what others have to say about this.
3. Once again, why should we go into those details if the definition is just unambiguous (not objective)? Not also sure why we should want to mechinism to be as small as possible. Having two buttons (play and pause) instead of one (that changes between play and pause) is as good as mechanism, right? Anyway, to try to mitigate that, I've added a note stating that rules must present an unambiguous description of the objective when using this definition. Can this work?

Reply:
1. ✔️ 
2. Hum, then I guess the question is who is the user 😉 "user" feels like "end user", i.e. the person browsing the page. Blocking access is more an objective for the dev than for the end user. But I do agree that it is maybe not needed to detail further.
3a. Two buttons (play and pause) can be one mechanism or two. And it feels currently ambiguous whether we consider them as one or two. Pushing that to the extreme, the full webpage (and all its scripting) is a single mechanism. Which is not really what's intended for the definition.
3b. I do like that. Rather than defining "mechanism" we're moving toward a definition of "mechanism to achieve an objective". I think this is a good direction to go as this remove a large part of the blurry and ambiguous intention behind the mechanism (typically for your 2 buttons example, this would disambiguate between "a mechanism to play and a mechanism to pause" (1 button each) and "a mechanism to play and pause" (both buttons)). Maybe we should do that move fully and actually defines "mechanism to achieve something"? What do you think? That could be a bit tricky on our side since the definition would have a "hole" (the "something" part), but that would enforce that this note is respected. Notably, the note is not really normative and is more a note to editors (rule writers) than a precision in the definition.

Reply:
2. I've always assumed the user in these rules to be the end-user but I've updated the definition to make that clear.
3. I like your suggestion to make this a definition of a "mechanism to achieve an objective". I've updated the rule name and description accordingly. Let's see what others think about this.